gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java

```diff
@@ -1063,26 +1092,41 @@ private void releaseQuota(DagNode<JobExecutionPlan> dagNode) {
       String proxyUser = ConfigUtils.getString(dagNode.getValue().getJobSpec().getConfig(), AzkabanProjectConfig.USER_TO_PROXY, null);
       if (proxyUser != null) {
         String proxyUserKey = DagManagerUtils.getUserQuotaKey(proxyUser, dagNode);
-        if (proxyUserToJobCount.containsKey(proxyUserKey) && proxyUserToJobCount.get(proxyUserKey) > 0) {
-          proxyUserToJobCount.put(proxyUserKey, proxyUserToJobCount.get(proxyUserKey) - 1);
-        }
+        decrementQuotaUsage(proxyUserToJobCount, proxyUserKey);
       }
 
       String serializedRequesters = DagManagerUtils.getSerializedRequesterList(dagNode);
       if (serializedRequesters != null) {
         try {
           for (ServiceRequester requester : RequesterService.deserialize(serializedRequesters)) {
             String requesterKey = DagManagerUtils.getUserQuotaKey(requester.getName(), dagNode);
-            if (requesterToJobCount.containsKey(requesterKey) && requesterToJobCount.get(requesterKey) > 0) {
-              requesterToJobCount.put(requesterKey, requesterToJobCount.get(requesterKey) - 1);
-            }
+            decrementQuotaUsage(requesterToJobCount, requesterKey);
           }
         } catch (IOException e) {
           log.error("Failed to release quota for requester list " + serializedRequesters, e);
         }
       }
     }
 
+    private void decrementQuotaUsage(Map<String, Integer> quotaMap, String user) {
+      Integer currentCount;
+      do {
+        currentCount = quotaMap.get(user);
+      } while (currentCount != null && currentCount > 0 && !quotaMap.replace(user, currentCount, currentCount - 1));
+    }
+
+    private void decrementQuotaUsageForUsers(Map<String, Integer> requesterAndCountToDecreaseMap) {
+      for (Map.Entry<String, Integer> entry : requesterAndCountToDecreaseMap.entrySet()) {
+        String user = entry.getKey();
+        Integer currentCount;
+        int newCount;
+        do {
+          currentCount = DagManagerThread.requesterToJobCount.get(user);
+          newCount = Math.max(0, DagManagerThread.requesterToJobCount.get(user) - entry.getValue());
+        } while (currentCount != null && currentCount > 0 && !DagManagerThread.requesterToJobCount.replace(user, currentCount, newCount));
```

Comment:
the short-circuit on both of these might not be solely that `currentCount > 0`, but that `currentCount != newCount`